### ISSUE TITLE
Global hotkey combinations with F-keys don't work for some users

### ISSUE THREAD

**Issue by lolmaus** [mintty/mintty#877]
Hi!

I've set up global hotkeys with Autohotkey using <kbd>F14</kbd> and <kbd>Shift</kbd>+<kbd>F14</kbd> key combinations (my keyboard does have <kbd>F14</kbd>).

Those key combinations work in every app except mintty (part of Git for Windows).

When I press <kbd>F14</kbd>, `~` is typed instead, and the action does not fire.
When I press <kbd>Shift</kbd>+<kbd>F14</kbd>, `;2~` is typed.

The Autohotkey action does not fire.

Pressing <kbd>F12</kbd> also types `~`. Pressing <kbd>F11</kbd> types `Function Key 11`.

Please help me resolve this.

**Comment by mintty** [comment]
In a terminal, function keys produce escape sequences so applications can read and recognize them.
On the shell command line, they are mostly not recognized, so after the ESC prefix, the rest of the code will be displayed, that's normal.
How a function key can be prevented from being recognized and swallowed by applications, so they would reach Autohotkey, I can only guess. For some testing, I'd need a reproducible test case, that is, a simple setup that will generate some feedback (e.g. pop up a message) when I type F14, or, say, F12.
As I'm not a user of Autohotkey and I don't have the spare time to study its lengthy documentation and it does not offer an intuitive configuration interface, I cannot setup a test case easily myself...

**Comment by Biswa96** [comment]
@lolmaus If you have a C compiler somewhere Try this gist https://gist.github.com/k-takata/4564ae4deca758420514. This can show the underlying code which may help to debug this issue.

**Comment by lolmaus** [comment]
@Biswa96 That program does not respond to <kbd>F14</kbd> regardless of whether Autohotkey is active.

For <kbd>F12</kbd>, it produces this:

```
Press key

Char received: Char code: 1b

Char received: [
Char code: 5b

Char received: 2
Char code: 32

Char received: 4
Char code: 34

Char received: ~
Char code: 7e
```

**Comment by mintty** [comment]
Please give me an Autohotkey setup (plug & play) that enables me to test this.

**Comment by mintty** [comment]
It's probably possible to define certain function keys explicitly as unused by mintty configuration so they will not be captured and still work outside.
But without a reproducible test case I cannot craft such a workaround so this issue shall be closed.

**Comment by lolmaus** [comment]
I'm travelling at the moment. I'll try to make a simple Autohotkey script when I'm back.

**Comment by mintty** [comment]
Downloaded Autohotkey and got it reported as Trojan by Windows Defender.
Sorry, I'm not going to test with it.
But you can assign a hotkey to a Windows shortcut, e.g. Control+F12.
If you do that, it works even when focussing in a mintty window, so I don't see a problem here.

**Comment by lolmaus** [comment]
You sure you got if from the right place?

No viruses are reported for the official distro: https://www.virustotal.com/gui/url/31dfe1dd5b9281a0908752c44df02d525f3d884c50e41fd24311e05d83273a96/detection

Regarding Windows shortcut, maybe an OS-level shortcut has a higher precedence? What if Autohotkey uses the same mechanism for scanning hotkeys as mintty and mintty wins?

**Comment by lolmaus** [comment]
It could be reported as trojan for listening to keypresses, that's what it's supposed to do.

**Comment by BrianInglis** [comment]
On Windows, OS level shortcuts have higher precedence, as Motif-derived ctrl/shift-insert/delete still work as copy/cut/paste work many places that application ctrl-C/X/V are not handled. 
Accelerator hot keys are supported in many apps and listening to keypresses does not make them trojans; even installing a application keyboard input monitoring hook would not normally be a problem. 
However installing a hook in the system wide keyboard input stack would be considered suspicion of keylogging, require whitelisting for programs such as AutoHotKey, and doing due diligence on where you download such programs, to ensure the input is not being logged and sent elsewhere. 
[Wikipedia](https://en.wikipedia.org/wiki/AutoHotkey) entries have good pointers to reliable download [sites](https://www.autohotkey.com/).

**Comment by mintty** [comment]
Installed AutoHotKey on another machine. But running it only opens a help file and offers no way to configure anything, so how is this beast supposed to work?

**Comment by lolmaus** [comment]
@mintty Thx for looking into this!

In order to use it you have to run an `*.ahk` script.

Here's a very basic script that fails for me in mintty (Git for Windows):

```ahk
#NoEnv  ; Recommended for performance and compatibility with future AutoHotkey releases.
#Warn  ; Enable warnings to assist with detecting common errors.
SendMode Input  ; Recommended for new scripts due to its superior speed and reliability.
SetWorkingDir %A_ScriptDir%  ; Ensures a consistent starting directory.

F12::
Send, My First Script
return
```

When I run it and press <kbd>F12</kbd> in any app, it would type `My First Script`.

Except mintty: when I press <kbd>F12</kbd>, it types `~`.

When I use <kbd>F11</kbd> instead of <kbd>F12</kbd> in mintty, it types `Function Key 11`.

The goal of this issue is to make my F-key combinations work in mintty: I use them to switch to specific keyboard layouts (rather than cycling layouts with Alt+Shift).

**Comment by lolmaus** [comment]
FYI: the only place other than mintty where my Autohotkey combinations do not work -- is the Windows start menu.

**Comment by mintty** [comment]
OK, so this thing works for me — it enters "My First Script" in mintty.
What's the issue?

**Comment by lolmaus** [comment]
It does not work for me. I have described the behavior above.

What can be causing the difference?

**Comment by mintty** [comment]
Closing as not reproducible and missing indication of a mintty issue.

**Comment by onemalt** [comment]
I have a very similar problem probably the same underlying cause. In my case I am using a program called Hot Keyboard but the problem of certain key combinations not being passed from mintty to Hot Keyboard sounds the same. I think I can "prove" this is a mintty issue, see my workaround below.

I have found a workaround though. I do _not_ have the problem if I revert to using version 'mintty 1.2-beta1 (x86_64-pc-cygwin)'. I _do_ have the problem when I use version 'mintty 3.0.2 (x86_64-pc-cygwin). I tried one other version of mintty in between those two versions and did not have the problem. But I don't own that computer anymore so I don't know which version it was.

I can certainly provide a lot more details about my Hot Keyboard setup and will be happy to do so if there is any interest. But in a nutshell, I highlight (select) some arbitrary text in a mintty window and then I press a key combination such as Alt-x which is received by Hot Keyboard, but when Hot Keyboard receives the Alt-x event it does not get the highlighted text put into its "Selection" variable. Instead, the Hot Keyboard "Selection" variable contains an empty text string, regardless of what text I had highlighted. The highlighted text is still in the clipboard as evidenced by the fact I can still paste it  with Ctrl-v.

**Comment by mintty** [comment]
OK, reopening. I am willing to look into the issue as soon as I can reproduce it.
Downloading Hot Keyboard Pro 4.0.25 which seems to have a 3-week test license, so the problem should be solved within 3 weeks :)

**Comment by mintty** [comment]
Hot Keyboard does not work on Windows 10, it shows the following error message:

---------------------------
Hot Keyboard error
---------------------------
Unfortunately, Windows 8 and Windows Server 2012 are not supported by this version of Hot Keyboard!
---------------------------
OK
---------------------------

**Comment by onemalt** [comment]
I have frequently used Hot Keyboard versions 4 and 6 on Windows 10. Specifically, I am running Hot Keyboard Pro version 6.5.112 right now on Windows 10 but I’ve previously run version 4.3.40 on the same Windows 10 computer. I’ve  managed to successfully avoid ever running Windows 8 and I’ve never run any server OS versions. I’m pretty sure this minty problem was happening with version 4.3.40 of Hot Keyboard too a few months ago but I’ve only now gotten around to debugging it and have upgraded to 6.5.112 in the meantime.

 

I’ll work on sending you a simple Hot Keyboard macro you can use to demonstrate the problem in the next hour.

 

Steve

**Comment by onemalt** [comment]
Attached are three simple Hot Keyboard macros. Inside Hot Keyboard, go to File->Import and import each of the .hk4backup files to the root menu. I also attached 4 PNG screen captures that show what Hot Keyboard should look like when the 3 test macros are correctly created. If the import doesn’t work for some reason you can pretty easily recreate the 3 test macros by looking at the PNG files that should show everything you need to know to create the macros manually.

 

If that works, then hitting your F10 function key should result in the text string “testing static test” being pasted. Simplest way to see that is open any application that receives text (e.g., Wordpad or a mintty window). Note that you will first need to minimize the Hot Keyboard window since the macros don’t work when the Hot Keyboard window is open.

 

If you make it this far then the next step is to confirm the other two macros work in a Wordpad/Notepad window as follows…

 

In Wordpad, type in some arbitrary text. Then use the mouse to highlight/select some of that text. Then hit Win-F9. That should cause the testSetCutBuffer9 macro to run and store your arbitrary text into a Hot Keyboard variable named ‘cutBuffer9’. Then still within Wordpad, hit F9 and the arbitrary text should be pasted. (The macro testPasteCutBuffer9 will simply paste the content of the ‘cutBuffer9’ variable, in whatever application is currently selected. I.e., you should be able to go to a mintty window and hit F9 and see your arbitrary text.

 

If all that works then try the preceding paragraph in a mintty window instead of Wordpad and if the mintty “bug” is present then the testSetCutBuffer9 macro will fail to copy your arbitrary text into the ‘cutBuffer9’ variable and instead seems to copy a null string which you can verify by hitting F9 and no text will be pasted. I typically to all this while running vim in a mintty window but you don’t actually need to use vim, you can demonstrate the problem straight from the mintty window. You can play around with those macros in things like Word and Excel other applications and they should work – it is only the mintty windows that don’t work.

 

Looking more carefully at the testSetCutBuffer9 macro in Hot Keyboard, the problem is this code: “SetVar ($cutBuffer9, Selection())” which calls Hot Keyboard’s built-in Selection() function which returns a text string representing the user’s currently “selection”, the same thing that normal Ctrl-C should copy into the clipboard. It is that Selection() function that returns an empty string. It’s certainly possible that Hot Keyboard is not doing something correctly, but I have demonstrated that simply changing my mintty executable to a different version affects the behavior.

 

Steve

 

From: Steve Herring [mailto:steve@raftalot.com] On Behalf Of github@raftalot.com
Sent: Thursday, April 02, 2020 9:13 AM
To: 'mintty/mintty'; 'mintty/mintty'
Cc: 'onemalt'; 'Comment'
Subject: RE: [mintty/mintty] Global hotkey combinations with F-keys don't work for some users (#877)

 

I have frequently used Hot Keyboard versions 4 and 6 on Windows 10. Specifically, I am running Hot Keyboard Pro version 6.5.112 right now on Windows 10 but I’ve previously run version 4.3.40 on the same Windows 10 computer. I’ve  managed to successfully avoid ever running Windows 8 and I’ve never run any server OS versions. I’m pretty sure this minty problem was happening with version 4.3.40 of Hot Keyboard too a few months ago but I’ve only now gotten around to debugging it and have upgraded to 6.5.112 in the meantime.

 

I’ll work on sending you a simple Hot Keyboard macro you can use to demonstrate the problem in the next hour.

 

Steve

**Comment by mintty** [comment]
Please upload the macros into the github issue.

**Comment by onemalt** [comment]
[testSetCutBuffer9.txt](https://github.com/mintty/mintty/files/4423775/testSetCutBuffer9.txt)
[testPasteCutBuffer9.txt](https://github.com/mintty/mintty/files/4423776/testPasteCutBuffer9.txt)
[testPasteCutBuffer10.txt](https://github.com/mintty/mintty/files/4423777/testPasteCutBuffer10.txt)

To import the 3 files into Hot Keyboard they must be renamed so the suffix is .hk4backup instead of .txt (Github has limitations on file types that can be uploaded).

**Comment by onemalt** [comment]
![hkb1](https://user-images.githubusercontent.com/63021876/78302018-e3234780-74ee-11ea-97ce-5745422dbec8.PNG)
![hkb2](https://user-images.githubusercontent.com/63021876/78302027-e6b6ce80-74ee-11ea-9817-395fff03dc6b.PNG)
![hkb3](https://user-images.githubusercontent.com/63021876/78302034-eae2ec00-74ee-11ea-9a68-f4d60b47644e.PNG)
![hkb4](https://user-images.githubusercontent.com/63021876/78302059-f6361780-74ee-11ea-86bc-cad56bf095a6.PNG)

These 4 PNG screen captures show what the Hot Keyboard macros should look like from the Hot Keyboard GUI. You could use these to create the macros manually if import doesn't work.

**Comment by mintty** [comment]
In response to #742, mintty clears its selection explicitly when another application modifies the clipboard. If this is disabled (winmain.c, when WM_CLIPBOARDUPDATE), the clipboard will persist. To be checked why precisely...
Whatever, there had been a previous change (96287d947fa32529829d9c8f655c8f8de27a8bd6) since which the first character of the hotkey-clipboard turn-around would be clipped by some weird interaction with keyboard handling. Tricky...

**Comment by mintty** [comment]
About the stripped first character, this was caused by Win+Ctrl injected by Hot Keyboard (for whatever reason...) which used to trigger the Compose key function when configured for Ctrl. Uploaded a patch not to trigger Compose when Win is held.

**Comment by mintty** [comment]
Fixed previous patch to a more generic guard against injected keys.

**Comment by mintty** [comment]
By analysis of messages delivered to mintty and resulting clipboard contents, Hot Keyboard works as follows with the two macros:

--------
Hot Keyboard copy macro

1. It clears the clipboard. This is notified to mintty (as configured) which in turn clears its own selection (it’s a feature, see #742).
2. Then it injects keyboard events Ctrl+KP_Insert to the application, to trigger the application to copy its selection to the clipboard.
It also injects additional Ctrl key events, possibly to neutralize the effect on other key events on the application but also possibly interfering with other application logic (such as the Compose function, for which I made a workaround).
3. Apparently it makes its own local copy from which the pasting macro will take its contents.

Notes:
* If the application does not copy on Ctrl+KP_Insert, this will not work.
* It’s unclear why Hot Keyboard initially clears the clipboard as it shall be overwritten by the application anyway. As this may effectively spoil the whole function (depending on the application if registered for a ClipboardFormatListener), please report this as an issue to Hot Keyboard. Mintty can add a workaround, disabling the clipboard listener but also disabling the clear-selection on other-selection feature.

--------
Hot Keyboard paste macro

1. It injects the pasted text by distinct key events. Unfortunately it uses the wrong modifier scancode for AltGr-composed characters. It uses 56 (LMENU) but it should use 312 (RMENU) combined with LCONTROL. This may cause wrong contents to be pasted, depending on the application.
2. It finally adds an unmotivated series of KEYUP events for various modifiers.

Notes:
* Please report the broken AltGr injection to Hot Keyboard. (A workaround for this is available in mintty already: set CtrlAltIsAltGr.)

**Comment by lolmaus** [comment]
Hi, topicstarter here.

Does any of this explain why I see `Function Key 11` typed into the terminal when I press <kbd>F11</kbd> in Git Bash?

* <kbd>F1</kbd>—<kbd>F4</kbd> — nothing (maybe they'll work with key combinations, haven't tried)
* <kbd>F5</kbd> — types `~`
* <kbd>F6</kbd>—<kbd>F11</kbd> — types `Function Key <N>`
* <kbd>F12</kbd> — types `~`
* <kbd>Shift</kbd>+<kbd>F1</kbd> — types `P`
* <kbd>Shift</kbd>+<kbd>F2</kbd> — types `Q`
* <kbd>Shift</kbd>+<kbd>F3</kbd> — types `R`
* <kbd>Shift</kbd>+<kbd>F4</kbd> — types `S`
* <kbd>Shift</kbd>+<kbd>F5</kbd> — types `;2~`
* <kbd>Shift</kbd>+<kbd>F6</kbd>—<kbd>F11</kbd> — types `2~`
* <kbd>Shift</kbd>+<kbd>F12</kbd> — types `;2~`

mintty 3.1.0 (I'm not sure if it's safe to update mintty apart from Git for Windows).

**Comment by mintty** [comment]
@onemalt, with the latest upload, Hot Keyboard is supported with `mintty -o CtrlAltIsAltGr=1 -o SupportExternalHotkeys=4`. Please report the two issues in my description above to that project.

**Comment by mintty** [comment]
@lolmaus, text inserted on F11 must be some test setup you have locally.
You can replace mintty in Git if you compile it in MSys2 (you could also add a development environment to Git but they made it deliberately hard to install it, seems to be a closed-shop deployment); you'd run mintty without Git patches then.

**Comment by lolmaus** [comment]
@mintty Uhm, I'm not aware of any test setup on my machine.

This behavior does not happen in any other app. E. g. <kbd>F11</kbd> toggles full screen in Chrome.

**Comment by mintty** [comment]
Maybe you put some KeyFunctions in your mintty config?

**Comment by lolmaus** [comment]
I didn't.

**Comment by mintty** [comment]
Or you have a laptop that needs Fn held for function keys...

**Comment by lolmaus** [comment]
It's a desktop, same problem with different keyboards, no Fn key.

And again, F-keys work as expected in all other apps.

**Comment by mintty** [comment]
The "Function Key" string is a specialty of Git Bash, not mintty.
Run `od -t x1` to verify mintty delivers the normal escape sequences.

**Comment by mintty** [comment]
They configured it in /etc/inputrc.

**Comment by lolmaus** [comment]
Thanks!

Feel free to close this issue or repurpose for the Hot Keyboard thing.

**Comment by onemalt** [comment]
I communicated the Hot Keyboard problems to those developers. Thanks much for the workaround in mintty.

**Comment by kde8192** [comment]
Hi!
I've been contacted by @onemalt regarding this issue and I represent Hot Keyboard support service.

> * It’s unclear why Hot Keyboard initially clears the clipboard as it shall be overwritten by the application anyway.

@mintty, unfortunately there is no standard way for a Windows app to get selected text from other Windows app since the second app can use some custom control which does not respond to a standard message like EM_GETSEL. So Hot Keyboard uses the following trick: it (indeed) clears the clipboard, simulates Ctrl-Ins and waits for some text to arrive. Most of the applications support Ctrl-Ins, so this method works and allows Hot Keyboard to get currently selected text. But you are completely right, if an app does not support Ctrl-Ins, Hot Keyboard won't be able to get any selected text from it.

Now why it clears the clipboard - just imagine that you already have some text in the clipboard before Hot Keyboard sends Ctrl-Ins to your application. Then how Hot Keyboard would know that your app put text into clipboard and that's not an old text that was there before? So, it clears the clipboard. Now if your app won't respond to Ctrl-Ins at all, after about 500 ms Hot Keyboard understands that no text is coming and stops waiting.

But I really don't understand why it's an issue - why your app monitors clipboard and does something on its cleaning?

> 2\. It also injects additional Ctrl key events, possibly to neutralize the effect on other key events on the application but also possibly interfering with other application logic

Yes. Hot Keyboard emulates modifier keys release in order for Ctrl-Ins to work. E.g. you assign Shift-1 to your macro. If hot keyboard just emulates Ctrl-Ins, you'll get Ctrl-Shift-Ins which is not what you want. So, it first emulates Shift release then sends Ctrl-Ins. Unfortunately if you assign Ctrl-1 to your macro, Hot Keyboard still sends Ctrl release at macro start since it may then need to switch the Num Lock off (and it does not respond to Ctrl-Num Lock). And yes, if your app does something on simple Ctrl click, it will be triggered. After Ctrl-Ins is sent, Hot Keyboard restores the original modifiers state.

> 1. It injects the pasted text by distinct key events. Unfortunately it uses the wrong modifier scancode for AltGr-composed characters. It uses 56 (LMENU) but it should use 312 (RMENU) combined with LCONTROL. This may cause wrong contents to be pasted, depending on the application.

Unfortunately Hot Keyboard does not support any additional characters in the play keys mode. It just uses Windows VkKeyScanEx function to convert symbol to its scan code and this function does not support AltGr. Also, I'm not sure how AtlGr works at all - it works exactly as an ordinary Alt key on my keyboard (I use English US layout).

> 2\. It finally adds an unmotivated series of KEYUP events for various modifiers.

Yes. This is done because the Paste Text macro in play keys mode uses the Record/Play macro engine which allows arbitrary keyboard and mouse events to be simulated. And this engine releases all modifier keys upon end of playback so no modifier key "stucks" independently on the sequence having been played.

**Comment by mintty** [comment]
@kde8192, thanks a lot for elaborating on this.
About clipboard clearing:

>But I really don't understand why it's an issue - why your app monitors clipboard and does something on its cleaning?

Let's consider 3 application instances A, B, C. Both A and B have selections, C wants to paste. Now for many apps (like notepad), there is no visual feedback which of A or B has copied the most recent selection and will be pasted. The user of C may not like that and prefer the older selection to be cleared in order to have unambiuous visual clipboard indication.
That is what the xterm terminal does and what was requested from mintty before, so I added monitoring the clipboard and clearing the selection on its change.

> Then how Hot Keyboard would know that your app put text into clipboard and that's not an old text that was there before? So, it clears the clipboard.

There don't seem to be timestamps on the Windows clipboard. But you could monitor the clipboard yourself and check whether you're getting a WM_CLIPBOARDUPDATE message after you inject Ctrl+Insert. I'd suggest to switch that non-destructive trick which would cooperate well with such application behaviour as I described.

About AltGr:

> It just uses Windows VkKeyScanEx function to convert symbol to its scan code and this function does not support AltGr.

You are right, this function, according to its doc, simplifies matters. It does not indicate left vs right modifier keys. However, the scancode it returns is assumedly that of the actual final key (e.g. Q), not of the modifiers. So you add your own assumption about modifier keycodes as a prefix within e.g. Ctrl, Alt, Q (which is @ on a German keyboard). As far as I know, Ctrl+Alt is never anything different from the AltGr assignment, so if you see this combination, you should simply replay the AltGr as it would come from the keyboard, meaning that you should send Alt with scancode 312 (which indicates right-Alt), not 56 as you do.

Thanks for caring.

**Comment by kde8192** [comment]
> Let's consider 3 application instances A, B, C. Both A and B have selections,

So, if I get you right - you have three instances of your app (AFAIU it's some kind of a terminal app), the user selects some text in A, copies it to clipboard, then it selects some text in B, copies it to clipboard and text selection in A disappears, right? So the users sees exactly one selection (in B) and knows that it will be pasted in C.

But what if the user selects some text in A, copies it to CB, then selects some text in B and leaves his PC for some time. Then he comes back and sees selection in A and B and now he is really confused which one will be pastes in C. Another interesting case is copying from third-party apps - there are no such selection-clean logic there, you can have a dozen of notepads with some text selected in each of them.

I use terminal clients from time to time myself and I am still confused about this feature - I often open a couple of terminal windows, select some text in A, then some text in B, then I can switch to A, copy to CB, switch to C, paste, switch to B, copy to CB, switch to C and paste again. Then I can repeat this sequence one more time and I am happy that my terminal client retains its selection.

> But you could monitor the clipboard yourself and check whether you're getting a WM_CLIPBOARDUPDATE

Thank you for the suggestion. We don't use this technique in Hot Keyboard now because it requires Windows Vista or later and Hot Keyboard is intended to run on WinXP. But it seems reasonable for me to make a conditional check - if AddClipboardFormatListener is available, Hot Keyboard will use it, otherwise it will retain the current logic.

But there is one more issue - when you ask Hot Keyboard to get selected text from an app, it does the following:

1. Saves current CB contents.
2. Clears CB.
3. Simulates Ctrl-Ins.
4. Waits for some text to arrive.
5. Get it.
6. Restores the original CB contents saved on step 1.

So, even if I remove step 2, it still will clear CB on step 6. It will allow HK to get selection from mintty, but mintty will still lose it on step 6. Is it OK?

> So you add your own assumption about modifier keycodes

Yes. And now HK always assumes the left versions of the modifier keys.

> As far as I know, Ctrl+Alt is never anything different from the AltGr assignment, so if you see this combination, you should simply replay the AltGr as it would come from the keyboard, meaning that you should send Alt with scancode 312 (which indicates right-Alt), not 56 as you do.

So, you mean that if VkKeyScanEx returns both Ctrl and Alt modifiers, HK should replace them by just one VK_RMENU? It can be easily done. But how can I test it myself? As I said before, I cannot make AtlGr on my keyboard produce any symbols, it works exactly like the ordinary left Alt key - e.g., if I click AltGr-F in Notepad, I get the "File" menu open.

**Comment by mintty** [comment]
About the selection scenarios, I forgot to mention (because it's the default I'm used to) that this is with auto-copy-on-select mode on, i.e. whenever you select some text in mintty, it's implicitly copied to the clipboard. So the visual indication reflects the actual clipboard contents, and it makes more sense to clear it if the clipboard changes than without that behaviour. I'll have to check on the implications for interworking if that option if off.

> 6. Restores the original CB contents saved on step 1.

This explains the second clipboard messages I saw, and some other behaviour I observed. Maybe you could skip this step if the current clipboard contents is the same as the previous one? That would catch the case in question.

----
> Yes. And now HK always assumes the left versions of the modifier keys.

That's exactly the problem.

> > As far as I know, Ctrl+Alt is never anything different from the AltGr assignment, so if you see this combination, you should simply replay the AltGr as it would come from the keyboard, meaning that you should send Alt with scancode 312 (which indicates right-Alt), not 56 as you do.

> So, you mean that if VkKeyScanEx returns both Ctrl and Alt modifiers, HK should replace them by just one VK_RMENU?

No, keep the VK_CONTROL and modify the VK_MENU to have scancode 312 (or use VK_RMENU, not sure whether that makes a difference).

> It can be easily done. But how can I test it myself? As I said before, I cannot make AtlGr on my keyboard produce any symbols, it works exactly like the ordinary left Alt key - e.g., if I click AltGr-F in Notepad, I get the "File" menu open.

You could install any international keyboard layout. Or I'll be happy to test for you, as long as I have a free license :)

**Comment by mintty** [comment]
Released 3.1.5.
